### PR TITLE
Polished PET-MAD-DOS example

### LINKED ISSUE #257: Polish PET-MAD-DOS example
The PET-MAD-DOS example works, but is very clunky. Some of the issues require making a better interface at the `upet` level, but I list them here as the end goal is to get a nice, user-friendly recipe.

- [x] bug (?): 
```
from upet.calculator import PETMADDOSCalculator
import ase-> Kernel crashed
import ase
from upet.calculator import PETMADDOSCalculator-> fine
```

- [x] I get a bunch of Warning: `You are sending unauthenticated requests to the HF Hub. Please set a HF_TOKEN to enable higher rate limits and faster downloads`, do we not cache the checkpoint locally?
- [x] In general, all the helper functions should be part of the upet.dos package. Doesn't make sense (and it's not sustainable) to have a bunch of opaque magic functions distributed with the recipe, e.g.:
    - computing the energy grid and/or interpolating on a different grid
    - computing the alignment
    - computing the structure DOS and mask
- [x] Have a utility function in upet that pads a given DOS profile
- [x] Decrease learning rate for fine-tuning
- [x] use the same energy grid for plotting true and predicted DOS (meaning true dos has to be padded, meaning we need better helper functions)
- [x] remove large-ish data/ files, use another zenodo
- [x] consider whether to use convolution to speed up calculation of the alignment loss (probably out of scope here)

### PR DESCRIPTION
**Pull request by HowWeiBin**
In response to the issues, I fixed the:
1) Import order
2) Integrated the helper functions in PETMADDOSCalculator and used them here
3) Decreased Learning rate to 1e-4
4) Used the same energy grid for both true and predicted DOS
5) Removed data files, entirely hosted on zenodo

### CODE REVIEW

**Comment by ceriottm** [comment]
ok, @pfebrer this is ready for you to check, merge and then all should be set to publish first version.

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 1]
```diff
@@ -43,217 +43,184 @@
...
+url = "https://zenodo.org/records/19655792/files/MAD_sample_structures.xyz?download=1"
+filename = "MAD_sample_structures.xyz"
+urllib.request.urlretrieve(url, filename)
+structs = ase.io.read("MAD_sample_structures.xyz", ":")
 
 # Extract the DOS and mask
-true_DOS = np.stack([s.info["DOS"] for s in structs])
-true_mask = np.stack([s.info["mask"] for s in structs])
+true_DOS = torch.tensor(np.stack([s.info["DOS"] for s in structs])).float()
+true_mask = torch.tensor(np.stack([s.info["mask"] for s in structs])).float()
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py]: Why? Did we have any problems with this?

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 2]
```diff
@@ -43,217 +43,184 @@
...
+# the computed DOS would indicate that there are no states past 3 eV. However, that is
+# false and is merely an artifact of eigenvalue truncation. Hence, an additional mask
+# is included for each structure to show the regions where the DOS is well-defined.
+# The DOS is considered well-defined up to 0.9eV below the minimum energy of the
+# highest band in the DFT calculation. Following, we plot an example DOS together
+# with its mask.
+#
 
-i_struct = 0  # index of the structure to visualize
+i_struct = 1  # index of the structure to visualize
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py:92]: Why?

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 3]
```diff
@@ -43,217 +43,184 @@
...
 
 # Plot DOS of the structure
-plt.plot(true_energy_grid, true_DOS[i_struct], label="DFT DOS", color="red")
+plt.plot(
+    pet_mad_dos_calculator._target_energy_grid,
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py]: We can't use private attributes in a tutorial, this needs to be made public somehow

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 4]
```diff
@@ -43,217 +43,184 @@
...
+    true_DOS[i_struct],
+    label="DFT DOS",
+    color="red",
+)
 # Plot mask for the structure (multiplied by 10 for better visualization)
 plt.plot(
-    true_energy_grid,
+    pet_mad_dos_calculator._target_energy_grid,
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py]: Same

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 5]
```diff
@@ -43,217 +43,184 @@
...
-# Here we will load the PET-MAD-DOS model using the upet package and use it to
-# predict the DOS. Although the target has size 4606 as seen in the cells above,
+# Although the target has size 4606 as seen in the cells above,
 # in order to accomodate the energy reference agnostic loss function,
-# PET-MAD-DOS predicts on a larger energy grid (`energies`) of size 4806.
-
-
-# Load the calculator
-pet_mad_dos_calculator = PETMADDOSCalculator(version="latest", device="cpu")
+# PET-MAD-DOS predicts on a larger energy grid (``energies``)of size 4806.
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py]: space after parenthesis

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 6]
```diff
@@ -43,217 +43,184 @@
...
 
 
 # %%
-# Step 2: Loading PET-MAD-DOS
+# Step 2: Predicting the DOS with PET-MAD-DOS
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-# Here we will load the PET-MAD-DOS model using the upet package and use it to
-# predict the DOS. Although the target has size 4606 as seen in the cells above,
+# Although the target has size 4606 as seen in the cells above,
 # in order to accomodate the energy reference agnostic loss function,
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py:126]: trailing comma should be removed

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 7]
```diff
@@ -43,217 +43,184 @@
...
-        adjusted_DOSes.append(torch.hstack([front_pad, target[index], end_pad]))
-        adjusted_mask.append(torch.hstack([front_pad, mask[index], end_pad]))
-    return torch.vstack(adjusted_DOSes), torch.vstack(adjusted_mask)
-
-
-# Adjust the DOS and mask based on the optimal shifts obtained above
-adjusted_true_DOS, adjusted_true_mask = adjust_DOS_and_mask(
-    torch.tensor(true_DOS), torch.tensor(true_mask), shifts
+denoised_DOS, aligned_true_DOS, aligned_true_masks = pet_mad_dos_calculator.align_dos(
+    denoised_pred_DOS, true_DOS.float(), true_mask.float()
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py]: Is it necessary to do the `.float()` here? Can't `align_dos` deal with different dtypes?

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 8]
```diff
@@ -285,134 +253,60 @@ def adjust_DOS_and_mask(target, mask, shift):
...
-        # Same smearing value as the one used in the MAD training data
-        smearing=0.3,
-        # For fine-tuning, we will use the same energy grid as the one used for
-        # training PET-MAD-DOS
-        energy_grid=true_energy_grid,
-        # PET-MAD-DOS predicts 200 more DOS
-        # values per structure (4806 - 4606 = 200)
-        n_extra_targets=200,
+    dos_i, mask_i = pet_mad_dos_calculator.compute_DOS_and_mask_from_eigenvalues(
+        torch.tensor(struct.info["eigvals"]), torch.tensor(struct.info["kweight"])
```
Comment (pfebrer) [inline examples/pet-mad-dos/pet-mad-dos.py]: Did people want this one hidden? I think the computation of the DOS could be kept in the tutorial, and then `PETMADDOSCalculator` could give you the mask, pad the DOS and so on

**Review by pfebrer — CHANGES_REQUESTED** [review]
Thanks for the changes, I have some comments. The most important one is that we can not use private attributes in tutorials.

File: examples/pet-mad-dos/pet-mad-dos.py [hunk 9]
```diff
@@ -285,134 +253,60 @@ def adjust_DOS_and_mask(target, mask, shift):
...
 # Store the processed structures as a new XYZ file for fine-tuning
 # ase.io.write("GaAs_processed_structures.xyz", GaAs_sample_structures)
 
 # %%
 # One would save the processed structures as a new XYZ file to to the
-# fine-tuning, as demonstrated in the commented line. However, for the
-# purposes of this tutorial, we will use datasets that have already been
-# processed, as can be obtained from the `MaterialsCloud archive
+# fine-tuning, as demonstrated in the commented line in the cell above. However,
+# for the purposes of this tutorial, we will demonstrate on datasets that have already
```
Comment (ceriottm) [inline examples/pet-mad-dos/pet-mad-dos.py]: Why do we need to download this if we have computed it? Either comment or (preferred) use the structure we just generated

**Review by ceriottm — CHANGES_REQUESTED** [review]
I'll do a few grammar changes directly, and add a chemiscope DOS plotting example.
PLS also update the function calls after converging on the API on the upet side, and if you use the model-specialized calls, add a note that there are generic versions in upet.utils.

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 9 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: I wanted to show that we need to pad the DOS if we use the data from the MaterialsCloud archive for fine-tuning. I didn't put the padded versions on MaterialsCloud because it might make it extra confusing if someone wants to train another foundation model from scratch or use the data in another other way.


I will include a section to explain it more explicitly

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 7 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: Removed

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 2 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py:92]: The denoising effect is clearer in Structure 1.

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 1 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: I remember there was some comment about how we should not have numpy arrays and torch tensors?

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 6 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py:126]: which one is the trailing comma? That sentence got restructured so maybe its not there anymore, please check this one again

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 8 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: I think the general comment is that we don't want functions in the notebook but to be integrated in upet

Comment (pfebrer) [reply examples/pet-mad-dos/pet-mad-dos.py]: Hmm ok I guess it's fine

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 1 above)
Comment (pfebrer) [reply examples/pet-mad-dos/pet-mad-dos.py]: Ok, do we need the `.float()`?

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 4 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: Done

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 3 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: Done

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 1 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: Removed!

File: examples/pet-mad-dos/pet-mad-dos.py (hunk 9 above)
Comment (HowWeiBin) [reply examples/pet-mad-dos/pet-mad-dos.py]: Done

**Comment by pfebrer** [comment]
Ok, so now the next steps are:
- Merge your `upet` changes into main.
- Make a `upet` release.
- Update this recipe to use the released `upet`.
- Merge this PR

**Comment by pfebrer** [comment]
Niceee the cookbook will work again :) Thanks!
